### ISSUE TITLE
Question: smart prompt

### ISSUE THREAD

**Issue by orefalo** [lmorg/murex#612]
Hi,

First and foremost, very nice shell - congrats - right on spot

coming from the fish shell, I miss one key feature -> a smart prompt such as [startfish](https://github.com/starship/starship)

Is there anything equivalent?
How would you bake "starfish like" features?

**Comment by lmorg** [comment]
@orefalo I've been very quiet on this issue. Just to let you know it hasn't been ignored. I have taken a look at starfish (and it looks really cool) and I think what I might do is submit a PR on their project to support this shell

**Comment by orefalo** [comment]
I already did ;-)

https://github.com/starship/starship/issues/5248

I started looking in more details. starfish is initialized via a setup command such as `starship init fish --print-full-init`

for fish, it generates

```
$ starship init fish --print-full-init
function fish_prompt
    switch "$fish_key_bindings"
        case fish_hybrid_key_bindings fish_vi_key_bindings
            set STARSHIP_KEYMAP "$fish_bind_mode"
        case '*'
            set STARSHIP_KEYMAP insert
    end
    set STARSHIP_CMD_PIPESTATUS $pipestatus
    set STARSHIP_CMD_STATUS $status
    # Account for changes in variable name between v2.7 and v3.0
    set STARSHIP_DURATION "$CMD_DURATION$cmd_duration"
    set STARSHIP_JOBS (count (jobs -p))
    if test "$TRANSIENT" = "1"
        # Clear from cursor to end of screen as `commandline -f repaint` does not do this
        # See https://github.com/fish-shell/fish-shell/issues/8418
        printf \e\[0J
        if type -q starship_transient_prompt_func
            starship_transient_prompt_func
        else
            printf "\e[1;32m❯\e[0m "
        end
    else
        /opt/homebrew/bin/starship prompt --terminal-width="$COLUMNS" --status=$STARSHIP_CMD_STATUS --pipestatus="$STARSHIP_CMD_PIPESTATUS" --keymap=$STARSHIP_KEYMAP --cmd-duration=$STARSHIP_DURATION --jobs=$STARSHIP_JOBS
    end
end

function fish_right_prompt
    switch "$fish_key_bindings"
        case fish_hybrid_key_bindings fish_vi_key_bindings
            set STARSHIP_KEYMAP "$fish_bind_mode"
        case '*'
            set STARSHIP_KEYMAP insert
    end
    set STARSHIP_CMD_PIPESTATUS $pipestatus
    set STARSHIP_CMD_STATUS $status
    # Account for changes in variable name between v2.7 and v3.0
    set STARSHIP_DURATION "$CMD_DURATION$cmd_duration"
    set STARSHIP_JOBS (count (jobs -p))
    if test "$TRANSIENT" = "1"
        if type -q starship_transient_rprompt_func
            starship_transient_rprompt_func
        else
            printf ""
        end
    else
        /opt/homebrew/bin/starship prompt --right --terminal-width="$COLUMNS" --status=$STARSHIP_CMD_STATUS --pipestatus="$STARSHIP_CMD_PIPESTATUS" --keymap=$STARSHIP_KEYMAP --cmd-duration=$STARSHIP_DURATION --jobs=$STARSHIP_JOBS
    end
end

# Disable virtualenv prompt, it breaks starship
set -g VIRTUAL_ENV_DISABLE_PROMPT 1

# Remove default mode prompt
builtin functions -e fish_mode_prompt

set -gx STARSHIP_SHELL "fish"

# Transience related functions
function reset-transient --on-event fish_postexec
    set -g TRANSIENT 0
end

function transient_execute
    if commandline --is-valid
        set -g TRANSIENT 1
        commandline -f repaint
    else
        set -g TRANSIENT 0
    end
    commandline -f execute
end

# --user is the default, but listed anyway to make it explicit.
function enable_transience --description 'enable transient prompt keybindings'
    bind --user \r transient_execute
    bind --user -M insert \r transient_execute
end

# Erase the transient prompt related key bindings.
# --user is the default, but listed anyway to make it explicit.
# Erasing a user binding will revert to the preset.
function disable_transience --description 'remove transient prompt keybindings'
    bind --user -e \r
    bind --user -M insert -e \r
end

# Set up the session key that will be used to store logs
# We don't use `random [min] [max]` because it is unavailable in older versions of fish shell
set -gx STARSHIP_SESSION_KEY (string sub -s1 -l16 (random)(random)(random)(random)(random)0000000000000000)
```

Didn't get time to go into the details yet,

also missing your side of the house - understanding how to configure murex prompt.

I need more time... ;-)

**Comment by lmorg** [comment]
oh good stuff.

There isn't an equivalent of `right_prompt` in Murex and currently the prompt doesn't support multiple lines. However one thing Murex does have that other shells do not is the hint text, The hint text sits below your prompt and offers context aware tips, like descriptions of commands and their parameters. You can think of it like a status bar actually.

You can configure the default hint text via `config`:
https://murex.rocks/docs/user-guide/interactive-shell.html#hint-text

Hint text doesn't support line breaks but long lines can wrap up to three lines deep, anything after that gets cropped.

**Comment by lmorg** [comment]
@orefalo I've also just added `pre-prompt-func` and `post-prompt-func` to `config`. You can define them in a similar way to the hint text instructions above. STDOUT and STDERR will write to the terminals STDOUT and STDERR, however STDIN will be ignored.

**Comment by orefalo** [comment]
thank you Laurence,
Will take it from there and write my first murex script ;-)

**Comment by orefalo** [comment]
I am now picking up on this item -  by now I grew sufficient murex's skills.

with that said.. I need help  ;-/

How would you implement a simple hellowworld_pre & post using `pre-prompt-func` and `post-prompt-func` in config?

I can't figure the proper `config` commands/section

thank you

**Comment by lmorg** [comment]
I removed the `pre-prompt-func` and `post-prompt-func` and turned them into events: https://dev.murex.rocks/events/onprompt.html#payload

So that means you're able to have triggers for `jump` and starfish without them conflicting with one another.

I think this may have been one of your many excellent suggestions actually 🤔

**Comment by orefalo** [comment]
that's right - I got distracted and missed the commit

**Comment by orefalo** [comment]
I am almost done, I read 3 other implementations for other shells.
The `onPrompt` event makes the murex implementation more elegant

With that said, here is what is missing (optional) so far:

* Murex displays the path at the beginning of the prompt, is there a way to customize?
* Other shells can read the terminal width with $COLUMNS, starship uses this information to shrten the prompt
* The ability to read intermediate errors when commands are pipes - $pipestatus  -> https://github.com/fish-shell/fish-shell/issues/2039
* This is just a suggestion, I believe the `rand str` alphabet should be restricted to `A-Za-z0-9_-` like nanoid does. That's because randoms are typically used in code and having hem generate special, non url friendly chars limits the usage. Or maybe you could add a parameter with different alphabet selections. default, A-Za-z0-9_-, A-Za-z..etc
* for optimization - the ability to generate the epoc time. Thinking about this one twice.. the time is used to calculate the command duration. Maybe murex could calculate it internally and pass it to the after event.

Hope this helps

**Comment by orefalo** [comment]
it's working

<img width="458" alt="CleanShot 2023-08-18 at 12 36 43" src="https://github.com/lmorg/murex/assets/52085/535c396b-5ad6-45d7-b87f-d95191607494">

**Comment by lmorg** [comment]
> The `onPrompt` event makes the murex implementation more elegant

That's good to know :)

> Murex displays the path at the beginning of the prompt, is there a way to customize?

That's managed via `config` (like onPrompt was originally intended).
```
» runtime --config -> [[ /shell ]] -> formap key val { if {$key =~ 'prompt'} {out $key: $val.Description} }
prompt: Interactive shell prompt
prompt-multiline: Shell prompt when command line string spans multiple lines

» config get shell prompt
{
    exitnum -> !if {
        out "{RESET}{YELLOW}${pwd_short -> grep -E -o '[^/]+$'} {GREEN}»{RESET} "
    } else {
        out "{RESET}{YELLOW}${pwd_short -> grep -E -o '[^/]+$'} {RED}»{RESET} "
    }
}

» config get shell prompt-multiline
{
    let len = ${pwd_short -> grep -E -o '[^/]+$' -> wc -c} - 1
    printf "%${$len}s » " $linenum
}
```

The idea of `prompt-multiline` is occasions where you have an incomplete command line. eg
![image](https://github.com/lmorg/murex/assets/12153701/2da3c40e-be0b-438d-8358-d73ba309a7c2)
So that will be unclosed brackets (eg `out {`), unclosed quotation marks, and, like the screenshot above, the last character being an escape.

> Other shells can read the terminal width with $COLUMNS, starship uses this information to shrten the prompt

Good suggestion. That would be easy to add

> The ability to read intermediate errors when commands are pipes - $pipestatus ->

I'd be interested to understand this requirement better. There are ways to have localised parts of the code that will fail the pipe mid pipe, eg

```
» err "fail" | out "foobar"
foobar
fail
dev » trypipe {err "fail" | out "foobar"}
fail
```
('fail' only appears after 'foobar' in the first example because STDOUT and STDERR are different threads (properly parallel jobs) and it takes ever so slightly longer to print STDERR)

It is also possible to query the state of commands when in test mode:
```
function example {
    test define bob %{
        ExitNum: 0
    }

    trypipe {
        err <test_bob> "fail" | out "foobar"
    }
}
```
![image](https://github.com/lmorg/murex/assets/12153701/44c20c67-6df1-4416-8c2a-7b9eeec53825)

As you know, I'm not against adding new features. But I would like to understand this particular use case better before committing any new code.

> This is just a suggestion, I believe the `rand str` alphabet should be restricted to `A-Za-z0-9_-` like nanoid does. That's because randoms are typically used in code and having hem generate special, non url friendly chars limits the usage. Or maybe you could add a parameter with different alphabet selections. default, A-Za-z0-9_-, A-Za-z..etc

I think I'd rather have a flag to add that behaviour. In the meantime, you might be able to use `rand int` to solve the problem:
```
» rand int
5234978402175678277
```

> for optimization - the ability to generate the epoc time. Thinking about this one twice.. the time is used to calculate the command duration. Maybe murex could calculate it internally and pass it to the after event.

There is already a builtin for timing blocks of code:
```
» time {sleep 5}
5.017743041
```
You could grab that with some named pipes:
```
pipe ptime                      # create a new named pipe
bg { <ptime> -> global vtime }  # background process to read from it
time <!ptime> { sleep 5 }       # time how long it takes to sleep, writing STDOUT to our new pipe
!pipe ptime                  # close the pipe
```
Though you'd have to wait a couple of seconds before querying `$GLOBAL.vtime` due to concurrency.

You can also use `datetime`: https://dev.murex.rocks/commands/datetime.html
```
» datetime --in "{now}" --out "{unix}"; echo; sleep 5; datetime --in "{now}" --out "{unix}"
1692355768
1692355773
```

**Comment by orefalo** [comment]
This is the ticket for the fish implementation -> https://github.com/fish-shell/fish-shell/issues/2039
it's a pretty good explanation of the problem statement.

datetime did the trick.

still fighting the prompt but getting there. thanks for the details

**Comment by lmorg** [comment]
What's the issue with the prompt?

**Comment by lmorg** [comment]
> This is the ticket for the fish implementation -> https://github.com/fish-shell/fish-shell/issues/2039
> it's a pretty good explanation of the problem statement.

Unless I've missed something, it doesn't explain the problem. Only a solution.

`trypipe` will return the exit number of the last command ran. eg
```
» trypipe {false | echo "foobar"}; exitnum
1
```
What I don't understand is what circumstances would there be for someone to want the pipeline to continue running on a non-zero exit number but still return the exit numbers of the commands that proceeded it? If the exit numbers matter than shouldn't you be escaping the pipeline after the first non-zero code?

**Comment by orefalo** [comment]
Correct, 

it's to check the exit status of ANY command in a pipeline, fish and bash both have it, and starfish can read the variable are properly report to the user.
https://www.shellscript.sh/examples/pipestatus/
if I got `trypipe` proper, it's more for scripts. the above is for the command line.
low priority for now to be honest,


regarding the prompt, I can't turn it completely off

```
murex » config set shell prompt { return " "  }
Invalid prompt. Block returned false.
murex » config set shell prompt { false  }
Invalid prompt. Block returned false.
murex » config set shell prompt { out ""  }
Invalid prompt. Block returned false.
...
```

**Comment by lmorg** [comment]
> if I got `trypipe` proper, it's more for scripts. the above is for the command line.

You can use `trypipe` in the command line too. Or `&&` (though that doesn't pipe)
![image](https://github.com/lmorg/murex/assets/12153701/25cebb2e-28d7-42d3-ab92-3e7c5e06c135)

You can force the interactive terminal (and shell scripts too) to run in `trypipe` mode via `-trypipe`. The only reason that isn't the default is because it breaks the ability for Murex to run each command in the pipe in parallel -- which normally doesn't matter but would definitely affect performance with super large data sets.

I did actually consider setting that as the default for v5. There have been a few users that have pushed for it to be the default too.

> regarding the prompt, I can't turn it completely off

Yeah, that's intentional. You could always go oldskool:
![image](https://github.com/lmorg/murex/assets/12153701/7d2a193a-1699-4e12-95c2-34b1fe0d06e3)
(the first error is because $PS1 wasn't set)

**Comment by orefalo** [comment]
ok - it's almost done..
but I am having an issue with the cursor.

<img width="567" alt="CleanShot 2023-08-20 at 16 25 19" src="https://github.com/lmorg/murex/assets/52085/c37c906f-7fa5-4abf-9f32-fe207b1d4d8b">

I think I need help...

**Comment by lmorg** [comment]
Sure. Though you'll need to share your code

**Comment by orefalo** [comment]
you may find the code here -> https://github.com/orefalo/murex-module-starship/blob/master/starship.mx

**Comment by lmorg** [comment]
Is starfish trying to create a multiline prompt? multiline prompts aren't currently supported. You can print additional lines in the **onPrompt before** event and just have the final line (the one the cursor sits on) as the **prompt** value.

**Comment by orefalo** [comment]
it does,
Unfortunately, starship does not provide a way to itemize line outputs.

**Comment by lmorg** [comment]
It doesn’t need to, murex can parse that easily. 

However hold off from making any changes just yet, I’ve thought up a really simple way to add multi line prompt support. A stupidly simple way (I wish it occurred to me before now!)

**Comment by orefalo** [comment]
I am also thinking about engaging with the starship team -
I read the doc last night, and they have obscure advanced settings which migh be of use

**Comment by lmorg** [comment]
Well I did find 5 minutes this evening so pushed out some experimental multi-line prompt support. it's in `develop` if you fancy giving that a try and seeing if it improves your starfish situation

**Comment by orefalo** [comment]
wow.. it's 98% working!
the only gap is the cursor.. which doesn't display one starship is used as the prompt.

<img width="418" alt="CleanShot 2023-08-24 at 17 38 33" src="https://github.com/lmorg/murex/assets/52085/b5647b65-681b-4215-b809-36b8762030fe">

it doesn't show on the picture but i can tell the cursor does print when ENTER is pressed.. maybe it gets overwritten by starship

**Comment by lmorg** [comment]
thanks for the module link. I'll import that and have a test

**Comment by lmorg** [comment]
sorry I'm not sure I understand the issue. this is what i see:
<img width="672" alt="image" src="https://github.com/lmorg/murex/assets/12153701/2fe8cc3b-66f7-441c-aa37-fa70681afe7b">

**Comment by orefalo** [comment]
Weird..
note how the cursor is blinking on the far right in me terminals

i tried with `iterm2` and the standard osx term

<img width="1064" alt="CleanShot 2023-08-25 at 00 31 00" src="https://github.com/lmorg/murex/assets/52085/f70bea4c-6329-4c8a-87af-04eb89122810">


I actually wonder why you don't have the issue...

any code you didn't push to develop yet?!

**Comment by orefalo** [comment]
try with my starship config. 

copy this file to ~/.config/starship.toml

```
# copy into ~/.config
# Replace the "❯" symbol in the prompt with "➜"
[character]      # The name of the module we are configuring is "character"
success_symbol = "➜ "
error_symbol = "✗ "

[kubernetes]
disabled = false
symbol = "☸ "
style = "bold blue"


format = """\
    $username\
    $hostname\
    $directory\
    $git_branch\
    $git_state\
    $git_status\
    $package\
    $golang\
    $java\
    $nodejs\
    $rust\
    $nix_shell\
    $aws\
    $env_var\
    $cmd_duration\
    $line_break\
    $jobs\
    $kubernetes\
    $character\
"""
```

**Comment by lmorg** [comment]
For some reason I cannot get your config to load. However I have spotted the problem, I forgot to add a "\r\n" after the prompt, so it kept overwriting the line.

As for why I didn't see the same problem, I'm running tmux so maybe that had something to do with it? The moment I launched Murex with the same starship module on the same terminal emulator but without tmux, I started seeing glitchy prompts.

I've just pushed a new patch to `develop`, give that a try
